app/Dockerfile

```diff
@@ -33,5 +64,76 @@ ENV HOST=0.0.0.0
 # Install application dependencies.
 # https://python-poetry.org/docs/basic-usage/#installing-dependencies
 RUN poetry install
+
 # Run the application.
 CMD ["poetry", "run", "python", "-m", "api"]
+
+#-------------------------------------------------------------------------------
+# Non-development build environment
+#-------------------------------------------------------------------------------
+FROM build as app-build
+
+# Runtime arguement for what user will the bulk of this stage run as
+ARG RUN_USER
+
+# In between ARG RUN_USER and USER ${RUN_USER}, the user is still root
+# If there is anything that needs to be ran as root, this is the spot
+
+# Change the user from root to ${RUN_USER}
+USER ${RUN_USER}
+
+# Set the application working directory.
+WORKDIR /app
+
+# install dependencies
+# use a project-local virtualenv to be easy to find
+ENV POETRY_VIRTUALENVS_IN_PROJECT true
+COPY pyproject.toml poetry.lock ./
+RUN poetry install --no-root --with app --without dev
```

Comment:
Can we add a comment for why we're specifically copying `pyproject.toml` and `poetry.lock` even though right after this we copy the entire `.` directory with `COPY . .`. My understanding is that functionally it's redundant, but that this way we can separate out the dependencies as an initial container layer to improve the cache performance of builds since dependencies don't change as often as the source code.

Reply:
Swapped to just using the `COPY . /app` since it covered everything

Reply:
Would there a performance gain to having it separate though? My understanding is that with the docker build cache, if the dependencies in pyproject.toml and poetry.lock don't change, Docker could reuse that layer, and only need to rebuild the layer with the source files?